File: ios/Sportivista/Profile/EntityPage.swift
//
//  EntityPage.swift
//  Sportivista
//
//  WP-170 — the ENTITY PAGE's pure core: «hva skjer med X?» answered in one
//  place, for one entity.
//
//  The go-to mapping (22.07) found this as the biggest remaining gap against
//  FotMob/VG Live: their core ritual is ONE page per team/athlete — next match,
//  last result, table position, news — and we did not have that object at all.
//  The answer was spread across Uka's agenda, the Nyheter board's four sections
//  and the event detail sheet.
//
//  This file is that object, and it is deliberately a COMPOSITION, not a new
//  data path. Every section is produced by machinery that already exists and is
//  already tested:
//
//    • KOMMENDE  — `FollowPresenter.nextEvents` (which is `FeedQuery.upcoming`
//                  under a single-rule lens). No new matching.
//    • RESULTAT  — `NewsBoard.resultRows` under a SINGLE-RULE `NewsLens`, so the
//                  rows, their per-sport DNA and — crucially — their
//                  `spoilerSensitive` flag are the SAME ones the Nyheter board
//                  and the WP-176 fulltidsvarsel compute. The spoiler shield is
//                  the existing `SpoilerShield`; this surface invents no second
//                  notion of "may I show an outcome".
//    • TABELL    — the WP-171 `StandingsTable` row type, filled by an ENTITY
//                  gate that is as honest as the event one: a table is shown
//                  ONLY when the entity is actually IN it. An OBOS club never
//                  gets the Premier League table.
//    • SISTE NYTT— `FollowPresenter.newsItems` (single-rule `NewsLens`).
//    • MER       — one deep link to the specialist, per VISJON v3 («ikke
//                  konkurrer på dybde, lenk til spesialisten»). Omitted, never
//                  faked, for a sport where we have no link we trust.
//
//  Everything here is Foundation-only and clock-injected, so EntityPageTests
//  drives every branch against the checked-in fixtures with no SwiftUI and no
//  running app — the same discipline as FollowPresenter/NewsBoard.
//

import Foundation

/// One KOMMENDE line on the entity page: the agenda's own answer (when · what ·
/// where), plus the full `Event` so a tap can open the SAME event detail sheet
/// the agenda opens.
struct EntityUpcomingRow: Identifiable {
    var id: String
    var title: String
    var sport: String
    /// «lør 25. · 18:00» for a single-day event, or the multi-day window as it
    /// already reads in the agenda.
    var whenLabel: String
    /// The first Norwegian channel, or an honest «–».
    var channelLabel: String
    var isMustSee: Bool
    /// The full event behind the row, when the cache still has it — a tap then
    /// opens the SAME detail sheet the agenda opens. `nil` degrades to a calm,
    /// non-tappable line rather than a dead-end tap.
    var event: Event?
}

/// A link OUT to the specialist for this sport. One row, at the bottom, never a
/// carousel — the entity page answers «hva skjer med X», it does not try to be
/// FotMob.
struct EntitySpecialistLink: Equatable {
    var label: String
    var url: URL
}

/// The whole page for one entity, in the FIXED section order the view renders:
/// anker · KOMMENDE · SISTE RESULTAT · TABELL · SISTE NYTT · MER.
struct EntityPage {
    var entity: Entity
    var upcoming: [EntityUpcomingRow]
    var results: [NewsResultRow]
    var table: StandingsTable?
    var news: [NewsItem]
    var specialist: EntitySpecialistLink?
    /// WP-30/WP-171 — does the user's spoiler policy cover THIS entity (or its
    /// sport)? The TABELL section is result-derived, so it hides behind the same
    /// «Vis tabell» reveal the event sheet uses when this is true. The per-result
    /// rows carry their own flag (`NewsResultRow.spoilerSensitive`), computed by
    /// the very same `SpoilerShield` — there is no second spoiler concept here.
    var spoilerSensitive: Bool = false

    /// Caps. Ro: this is ONE screen with a fixed set of sections, never an
    /// infinite scroll. Each section says the little it can stand behind.
    static let upcomingLimit = 3
    static let resultLimit = 3
    static let newsLimit = 4
    /// Rows a table shows before the entity's own row is appended — the WP-171
    /// value, so the event sheet and this page read identically.
    static let tableTopRows = StandingsTable.topRows

    /// True when we know NOTHING about this entity right now. The view then
    /// shows ONE honest line instead of four empty section shells.
    var isEmpty: Bool { upcoming.isEmpty && results.isEmpty && table == nil && news.isEmpty }

    // MARK: - Build

    /// Compose the page. `rule` is the follow rule when the page was reached
    /// from «Det du følger» (so a WHOLE-SPORT follow keeps its whole-sport
    /// semantics); from an event detail sheet there is no rule and a stand-in is
    /// derived from the entity itself.
    static func build(
        entity: Entity,
        rule: InterestRule? = nil,
        feed: FeedQuery,
        index: EntityIndex,
        eventsById: [String: Event] = [:],
        news: [NewsItem] = [],
        results: RecentResults = RecentResults(),
        standings: Standings? = nil,
        tracked: TrackedConfig? = nil,
        shield: SpoilerShield = SpoilerShield(),
        now: Date = Date()
    ) -> EntityPage {
        let effectiveRule = rule ?? standInRule(for: entity)
        let presenter = FollowPresenter(feed: feed, index: index, news: news, tracked: tracked, now: now)
        let upcomingRows = presenter.nextEvents(for: effectiveRule, limit: upcomingLimit)

        // The SINGLE-RULE lens: exactly the same lens the Nyheter board uses,
        // narrowed to this one entity — so a result is on this page for the same
        // reason (and with the same spoiler verdict) it would be on the board.
        let lens = NewsLens(profile: InterestProfile(rules: [effectiveRule]), index: index)
        let resultRows = lens.isEmpty
            ? []
            : Array(NewsBoard.resultRows(results, lens: lens, index: index, shield: shield).prefix(resultLimit))

        return EntityPage(
            entity: entity,
            upcoming: upcomingRows.map { row(for: $0, presenter: presenter, eventsById: eventsById) },
            results: resultRows,
            table: standingsTable(entity: entity, standings: standings),
            news: presenter.newsItems(for: effectiveRule, limit: newsLimit),
            specialist: specialistLink(sport: entity.sport.isEmpty ? effectiveRule.sport : entity.sport, name: entity.name),
            spoilerSensitive: shield.isSpoilerSensitive(
                sport: NewsLens.canonicalSport(entity.sport.isEmpty ? effectiveRule.sport : entity.sport),
                entityIds: [entity.id]
            )
        )
    }

    /// A stand-in follow rule for an entity we did NOT arrive at through a
    /// follow (the event-detail entry point). It carries only the identity —
    /// the matching downstream is entirely id/name based, exactly as for a real
    /// rule, so the two entry points produce the same page.
    static func standInRule(for entity: Entity) -> InterestRule {
        InterestRule(
            entityId: entity.id,
            entityName: entity.name,
            sport: entity.sport,
            weight: InterestProfile.defaultWeight,
            reason: "",
            addedAt: Date(timeIntervalSince1970: 0)
        )
    }

    /// Project a feed row onto the page's KOMMENDE line, attaching the full
    /// `Event` when the cache still carries it (so a tap opens the normal detail
    /// sheet).
    private static func row(for event: FeedQueryEvent, presenter: FollowPresenter, eventsById: [String: Event]) -> EntityUpcomingRow {
        EntityUpcomingRow(
            id: event.id,
            title: event.title,
            sport: event.sport,
            whenLabel: whenLabel(for: event, presenter: presenter),
            channelLabel: event.channelLabel,
            isMustSee: event.isMustSee,
            event: eventsById[event.id]
        )
    }

    /// «lør 25. · 18:00»; a multi-day window (already a date range) stands on
    /// its own. Same shape as the follow detail's KOMMENDE line — one voice.
    static func whenLabel(for event: FeedQueryEvent, presenter: FollowPresenter) -> String {
        if event.timeLabel.contains("–") { return event.timeLabel }
        return "\(presenter.shortDayLabel(dayKey: event.dayKey)) · \(event.timeLabel)"
    }

    // MARK: - TABELL (the honest entity gate)

    /// The table this ENTITY is actually in — never «a» table for its sport.
    /// WP-171's lesson, applied to the entity: an OBOS-league club must not be
    /// shown the Premier League top five as if it were its own table, and a
    /// golfer who isn't in this week's field has no leaderboard. `nil` whenever
    /// the entity is not found in any table we publish (the view then omits the
    /// section entirely — never an empty shell).
    static func standingsTable(entity: Entity, standings: Standings?) -> StandingsTable? {
        guard let standings, !standings.isEmpty else { return nil }
        let terms = ([entity.name] + entity.aliases).filter { !$0.isEmpty }
        guard !terms.isEmpty else { return nil }
        let hit: (String) -> Bool = { name in
            guard !name.isEmpty else { return false }
            return terms.contains { TextMatch.containsName(name, $0) || TextMatch.containsName($0, name) }
        }

        switch TextMatch.normalize(entity.sport) {
        case "football":
            for (title, table) in [("PREMIER LEAGUE", standings.football.premierLeague), ("LA LIGA", standings.football.laLiga)] {
                guard table.contains(where: { hit($0.team) }) else { continue }
                var picked = Array(table.prefix(tableTopRows))
                for entry in table where hit(entry.team) && !picked.contains(where: { $0.position == entry.position }) {
                    picked.append(entry)
                }
                picked.sort { $0.position < $1.position }
                return StandingsTable(title: title, rows: picked.map { entry in
                    StandingsTableRow(
                        id: "table|\(entry.position)|\(entry.team)",
                        rank: "\(entry.position).",
                        name: entry.team,
                        value: "\(entry.points)",
                        highlighted: hit(entry.team)
                    )
                })
            }
            return nil

        case "golf":
            for key in standings.golf.keys.sorted() {
                guard let board = standings.golf[key], !board.leaderboard.isEmpty else { continue }
                let all = board.leaderboard + board.trackedPlayers
                guard all.contains(where: { hit($0.player) }) else { continue }
                var rows = board.leaderboard.prefix(tableTopRows).map { golfRow($0, highlighted: hit($0.player)) }
                for entry in all where hit(entry.player) && !rows.contains(where: { $0.name == entry.player }) {
                    rows.append(golfRow(entry, highlighted: true))
                }
                return StandingsTable(title: (board.name ?? "LEDERTAVLE").uppercased(), rows: rows)
            }
            return nil

        case "f1":
            let drivers = standings.f1.drivers
            guard drivers.contains(where: { hit($0.driver) }) else { return nil }
            var picked = Array(drivers.prefix(tableTopRows))
            for entry in drivers where hit(entry.driver) && !picked.contains(where: { $0.position == entry.position }) {
                picked.append(entry)
            }
            picked.sort { $0.position < $1.position }
            return StandingsTable(title: "VM-STILLING", rows: picked.map { d in
                StandingsTableRow(
                    id: "f1|\(d.driver)",
                    rank: "\(d.position).",
                    name: d.driver,
                    value: "\(d.points)",
                    highlighted: hit(d.driver)
                )
            })

        default:
            return nil
        }
    }

    private static func golfRow(_ entry: GolfStandingRow, highlighted: Bool) -> StandingsTableRow {
        StandingsTableRow(
            id: "golf|\(entry.player)",
            rank: entry.positionDisplay ?? entry.position.map { "\($0)." } ?? "–",
            name: entry.player,
            value: entry.score ?? "–",
            highlighted: highlighted
        )
    }

    // MARK: - MER (the deep link out)

    /// The specialist we hand the user off to for DEPTH — squad lists, xG, lap
    /// charts. VISJON v3: we do not compete on depth, we link to the one who
    /// owns it.
    ///
    /// Deliberately NARROW. A link that 404s or lands on a front page is worse
    /// than no link at all, so the map holds only search endpoints we are
    /// confident about, and every other sport simply has no MER section. Growing
    /// it is a one-line change once a URL has been verified by hand.
    static func specialistLink(sport: String, name: String) -> EntitySpecialistLink? {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        guard let query = trimmed.addingPercentEncoding(withAllowedCharacters: .urlQueryValueAllowed) else { return nil }
        let template: (label: String, prefix: String)
        switch NewsLens.canonicalSport(sport) {
        case "football":
            template = ("FotMob", "https://www.fotmob.com/search?term=")
        case "cycling":
            template = ("ProCyclingStats", "https://www.procyclingstats.com/search.php?term=")
        case "esports":
            template = ("Liquipedia", "https://liquipedia.net/counterstrike/index.php?search=")
        default:
            return nil
        }
        guard let url = URL(string: template.prefix + query) else { return nil }
        return EntitySpecialistLink(label: template.label, url: url)
    }
}

private extension CharacterSet {
    /// Query-VALUE safe: `urlQueryAllowed` still permits `&`, `+` and `=`, which
    /// would break a name like "Brann & Co" out of its own parameter.
    static let urlQueryValueAllowed = CharacterSet.urlQueryAllowed.subtracting(CharacterSet(charactersIn: "&+=?"))
}

File: docs/js/entity-page.js
// entity-page.js — WP-170: the ENTITY PAGE, «hva skjer med X?» in one place.
// The web twin of ios/Sportivista/Profile/EntityPage.swift + EntityPageView.swift.
//
// The go-to mapping (22.07) found this as the biggest remaining gap against
// FotMob/VG Live: their core ritual is ONE page per team/athlete — next match,
// last result, table position, news — and we had no such object at all. The
// answer was spread across the agenda, the Nyheter board's sections and the
// event detail.
//
// It is a COMPOSITION of what the page already loaded — no new data file, no new
// request, no new matching:
//   KOMMENDE       — the SAME `ssNextEventForEntity` term matching «Dette dekker
//                    vi» uses, widened to the next few events.
//   SISTE RESULTAT — `ssResultRows(recent-results)` (news-web.js), name-filtered.
//   TABELL         — an ENTITY gate over standings.json that is as honest as the
//                    event one: the table is shown ONLY when the entity is in it.
//                    An OBOS club never gets the Premier League table.
//   SISTE NYTT     — news.json pointers matched on entityIds, else on the name.
//   MER            — one deep link to the specialist (VISJON v3: «ikke konkurrer
//                    på dybde, lenk til spesialisten»). Absent, never guessed.
//
// Calm rules (they are what keep this from becoming a FotMob clone): ONE screen,
// a FIXED section order, no tabs, no infinite scroll, small caps per section, and
// every section OMITTED when it has nothing to say.
//
// NB: the web has no spoiler shield — `spoilerPolicy` is an on-device iOS memory
// fact and never leaves the phone. So the web shows results plainly here, exactly
// as the Nyheter board already does (news-web.js `resultRow`). The iOS page keeps
// the shield.

/** How much each section shows. Ro before completeness. */
const SS_EP_UPCOMING = 3;
const SS_EP_RESULTS = 3;
const SS_EP_NEWS = 4;
/** Table rows before the entity's own row is appended (mirrors StandingsTable.topRows). */
const SS_EP_TABLE_TOP = 5;

/** The match terms for an entity entry (a bare name, an alwaysTrack entry, or an
 *  entities.json record) — name + aliases, exactly what the rest of the board
 *  matches on. */
function ssEntityTerms(entry) {
	return trackedTerms([entry]).filter(Boolean);
}

/** Does any of the entity's terms hit this haystack? Word-boundary + accent
 *  insensitive (`ssContainsTerm`), never a substring free-for-all. */
function ssEntityHits(haystack, entry) {
	const hay = String(haystack || '');
	return ssEntityTerms(entry).some((t) => ssContainsTerm(hay, t));
}

/** The next `limit` upcoming events for an entity, nearest first. Same sport
 *  scoping and same 3h "already over" floor as `ssNextEventForEntity` — it is
 *  literally applied repeatedly over a shrinking pool, so the first row here is
 *  ALWAYS the row «Dette dekker vi» shows. */
function ssEntityUpcoming(events, entry, now = Date.now(), limit = SS_EP_UPCOMING) {
	const out = [];
	let pool = Array.isArray(events) ? events.slice() : [];
	for (let i = 0; i < limit; i++) {
		const next = ssNextEventForEntity(pool, entry, now);
		if (!next) break;
		out.push(next);
		pool = pool.filter((e) => e !== next);
	}
	return out;
}

/** The recent results this entity took part in (rows from news-web.js's
 *  `ssResultRows`, whose `names` carry the participants). */
function ssEntityResults(rows, entry, limit = SS_EP_RESULTS) {
	return (rows || [])
		.filter((r) => ssEntityHits((r.names || []).join(' '), entry))
		.slice(0, limit);
}

/** The news pointers about this entity: a stamped entityId hit when we know the
 *  id, else the name in the headline. Never the whole sport — this is a page
 *  about ONE entity. */
function ssEntityNews(news, entry, entityId, limit = SS_EP_NEWS) {
	const id = entityId || (entry && entry.id) || null;
	return (news || [])
		.filter((it) => {
			const ids = Array.isArray(it.entityIds) ? it.entityIds : [];
			if (id && ids.indexOf(id) >= 0) return true;
			return ssEntityHits(it.title, entry);
		})
		.slice()
		.sort((a, b) => (Date.parse(b.publishedAt || 0) - Date.parse(a.publishedAt || 0)))
		.slice(0, limit);
}

/**
 * The table this ENTITY is actually in — never «a» table for its sport.
 * Returns `{ title, rows: [{ rank, name, value, highlighted }] }` or null.
 *
 * The honest gate is the whole point (the WP-171 lesson, applied per entity):
 * football events cover far more leagues than we publish standings for, so an
 * OBOS fixture must not render the Premier League top five as if it were its own
 * table. Twin of `EntityPage.standingsTable` in Swift.
 */
function ssEntityStandingsTable(standings, entry) {
	if (!standings) return null;
	const sport = (entry && entry.sport) || '';
	const hit = (name) => ssEntityHits(name, entry);

	if (sport === 'football') {
		const tables = standings.football || {};
		for (const [title, table] of [['Premier League', tables.premierLeague], ['La Liga', tables.laLiga]]) {
			if (!Array.isArray(table) || !table.length) continue;
			if (!table.some((t) => hit(t.team))) continue;
			const picked = table.slice(0, SS_EP_TABLE_TOP);
			for (const t of table) if (hit(t.team) && !picked.some((p) => p.position === t.position)) picked.push(t);
			picked.sort((a, b) => a.position - b.position);
			return { title, rows: picked.map((t) => ({ rank: `${t.position}.`, name: t.team, value: String(t.points), highlighted: hit(t.team) })) };
		}
		return null;
	}

	if (sport === 'golf') {
		const tours = standings.golf || {};
		for (const key of Object.keys(tours).sort()) {
			const board = tours[key];
			const lb = (board && Array.isArray(board.leaderboard)) ? board.leaderboard : [];
			const tracked = (board && Array.isArray(board.trackedPlayers)) ? board.trackedPlayers : [];
			if (!lb.length) continue;
			if (!lb.concat(tracked).some((p) => hit(p.player))) continue;
			const rows = lb.slice(0, SS_EP_TABLE_TOP).map((p) => ssGolfTableRow(p, hit(p.player)));
			for (const p of lb.concat(tracked)) {
				if (hit(p.player) && !rows.some((r) => r.name === p.player)) rows.push(ssGolfTableRow(p, true));
			}
			return { title: board.name || 'Ledertavle', rows };
		}
		return null;
	}

	if (sport === 'f1') {
		const drivers = (standings.f1 && Array.isArray(standings.f1.drivers)) ? standings.f1.drivers : [];
		if (!drivers.some((d) => hit(d.driver))) return null;
		const picked = drivers.slice(0, SS_EP_TABLE_TOP);
		for (const d of drivers) if (hit(d.driver) && !picked.some((p) => p.position === d.position)) picked.push(d);
		picked.sort((a, b) => a.position - b.position);
		return { title: 'VM-stilling', rows: picked.map((d) => ({ rank: `${d.position}.`, name: d.driver, value: String(d.points), highlighted: hit(d.driver) })) };
	}

	return null;
}

function ssGolfTableRow(p, highlighted) {
	return {
		rank: p.positionDisplay || (p.position != null ? `${p.position}.` : '–'),
		name: p.player,
		value: p.score || '–',
		highlighted,
	};
}

/**
 * The specialist we hand the reader off to for DEPTH (squad lists, xG, lap
 * charts). VISJON v3: we don't compete on depth, we link to the one who owns it.
 *
 * Deliberately NARROW, and identical to the Swift map: a link that 404s or lands
 * on a front page is worse than no link, so only search endpoints we are
 * confident about are here; every other sport simply has no MER section.
 */
function ssSpecialistLink(sport, name) {
	const who = String(name || '').trim();
	if (!who) return null;
	const table = {
		football: { label: 'FotMob', prefix: 'https://www.fotmob.com/search?term=' },
		cycling: { label: 'ProCyclingStats', prefix: 'https://www.procyclingstats.com/search.php?term=' },
		esports: { label: 'Liquipedia', prefix: 'https://liquipedia.net/counterstrike/index.php?search=' },
	};
	const t = table[typeof ssCanonicalNewsSport === 'function' ? ssCanonicalNewsSport(sport) : String(sport || '')];
	if (!t) return null;
	return { label: t.label, url: t.prefix + encodeURIComponent(who) };
}

// ── Dashboard prototype extension (DOM) ──────────────────────────────────────
if (typeof Dashboard !== 'undefined') Object.assign(Dashboard.prototype, {

	/** Remember an entry so a row can reopen it by key (the row markup carries a
	 *  key, never a serialised object). */
	rememberEntity(entry) {
		const key = ssNormalize(ssEntityName(entry)).trim();
		if (!key) return '';
		this._entityEntries = this._entityEntries || new Map();
		if (!this._entityEntries.has(key)) this._entityEntries.set(key, entry);
		return key;
	},

	entityForKey(key) {
		return (this._entityEntries && this._entityEntries.get(key)) || null;
	},

	/** The registry record behind an entry, when entities.json knows it — the
	 *  source of the avatar (WP-185/186) and of the stable id for news matching. */
	entityRecord(entry) {
		if (typeof this.identityIndex !== 'function') return null;
		const { byName } = this.identityIndex();
		return byName.get(ssNormalize(ssEntityName(entry)).trim()) || null;
	},

	/** The whole page for one entity, as HTML. Pure-ish (reads this.allEvents /
	 *  standings / recentResults / news) so a test can assert the composed page
	 *  without a DOM. */
	entityPageHtml(entry) {
		const name = ssEntityName(entry);
		const record = this.entityRecord(entry);
		const sport = (entry && entry.sport) || (record && record.sport) || '';
		const avatar = (typeof ssEntityIdentity === 'function' && typeof ssEntityAvatar === 'function')
			? ssEntityAvatar(ssEntityIdentity(record))
			: '';
		const sub = [this.sportLabel ? this.sportLabel(sport) : sport, record && record.type ? SS_EP_TYPE_NB[record.type] || '' : '']
			.filter(Boolean).join(' · ');

		let html = `<div class="ep-head">
			<span class="ep-avatar" aria-hidden="true">${avatar || (this.sportIconCell ? this.sportIconCell(sport) : '')}</span>
			<span class="ep-id"><span class="ep-name">${escapeHtml(name)}</span>${sub ? `<span class="ep-sub">${escapeHtml(sub)}</span>` : ''}</span>
			<button type="button" class="sheet-close" data-entity-close>Lukk</button>
		</div>`;

		let any = false;

		// 1. KOMMENDE — the next few events, the first one expanded with the
		//    når·hva·hvor detail (incl. the golfer's own tee time via followDetail).
		const upcoming = ssEntityUpcoming(this.allEvents, entry);
		if (upcoming.length) {
			any = true;
			html += `<section class="ep-section"><h3 class="ep-h">Kommende</h3>`
				+ `<div class="ep-detail">${this.followDetail(upcoming[0], entry)}</div>`
				+ upcoming.slice(1).map((e) => this.entityUpcomingRow(e)).join('')
				+ `</section>`;
		}

		// 2. SISTE RESULTAT — the same rows the Nyheter board renders.
		const results = ssEntityResults(typeof ssResultRows === 'function' ? ssResultRows(this.recentResults) : [], entry);
		if (results.length) {
			any = true;
			html += `<section class="ep-section"><h3 class="ep-h">Siste resultat</h3>${results.map((r) => this.resultRow(r)).join('')}</section>`;
		}

		// 3. TABELL — only when it is genuinely this entity's table.
		const table = ssEntityStandingsTable(this.standings, Object.assign({}, typeof entry === 'string' ? { name: entry } : entry, { sport }));
		if (table && table.rows.length) {
			any = true;
			html += `<section class="ep-section"><h3 class="ep-h">${escapeHtml(table.title)}</h3>`
				+ table.rows.map((r) => `<div class="ep-table-row${r.highlighted ? ' is-entity' : ''}">`
					+ `<span class="ep-rank">${escapeHtml(r.rank)}</span>`
					+ `<span class="ep-team">${escapeHtml(r.name)}</span>`
					+ `<span class="ep-val">${escapeHtml(r.value)}</span></div>`).join('')
				+ `</section>`;
		}

		// 4. SISTE NYTT — pointers to the source, never our own summary.
		const news = ssEntityNews(this.news, entry, record && record.id);
		if (news.length) {
			any = true;
			html += `<section class="ep-section"><h3 class="ep-h">Siste nytt</h3>${news.map((it) => this.newsRow(it)).join('')}</section>`;
		}

		// 5. MER — the hand-off. Never a guessed URL.
		const link = ssSpecialistLink(sport, name);
		if (link) {
			html += `<section class="ep-section"><h3 class="ep-h">Mer</h3>`
				+ `<a class="ep-link" href="${escapeHtml(link.url)}" target="_blank" rel="noopener">Åpne i ${escapeHtml(link.label)} ↗</a>`
				+ `<p class="ep-note">Vi svarer på når og hvor. Dybden — tropp, statistikk — er andres fag.</p></section>`;
		}

		if (!any) {
			html += `<p class="ep-empty">Ingenting på tavla om ${escapeHtml(name)} akkurat nå. Siden fyller seg når det kommer events, resultater eller nyheter.</p>`;
		}
		return html;
	},

	/** One quiet extra KOMMENDE line (the first event is shown expanded above). */
	entityUpcomingRow(e) {
		const d = new Date(e.time);
		const when = `${d.toLocaleDateString('nb-NO', { weekday: 'short', day: 'numeric', month: 'short', timeZone: 'Europe/Oslo' })} ${this.osloTime(d)}`;
		const what = (e.homeTeam && e.awayTeam) ? `${ssShortName(e.homeTeam)} – ${ssShortName(e.awayTeam)}` : (e.title || '');
		const chan = (Array.isArray(e.streaming) && e.streaming[0] && (e.streaming[0].platform || e.streaming[0])) || '';
		return `<div class="ep-row"><span class="ep-when">${escapeHtml(when)}</span>`
			+ `<span class="ep-what">${escapeHtml(what)}</span>`
			+ `<span class="ep-where">${chan ? escapeHtml(String(chan)) : '<span class="tbd">–</span>'}</span></div>`;
	},

	/** The event-detail links that lead TO the entity page (WP-170): one per side
	 *  this event is about (each team, each Norwegian player). Reuses
	 *  `followTargets` so it resolves the SAME real entities.json ids the «Følg X»
	 *  buttons use — so the page's news matching keys on the stable id. Empty (no
	 *  section) when nothing resolves — the honest degradation. */
	entityActionsHtml(e) {
		if (typeof this.followTargets !== 'function') return '';
		const targets = this.followTargets(e);
		if (!targets.length) return '';
		const links = targets.map((t) => {
			const entry = { id: t.entityId, name: t.entityName, sport: t.sport, aliases: [] };
			const key = this.rememberEntity ? this.rememberEntity(entry) : '';
			return `<button type="button" class="ev-act ev-entity" data-entity-key="${escapeHtml(key)}">${escapeHtml(t.entityName)} →</button>`;
		}).join('');
		return `<span class="d-entities">${links}</span>`;
	},

	openEntityPage(entry) {
		const wrap = document.getElementById('entity-page');
		const body = document.getElementById('entity-page-body');
		if (!wrap || !body || !entry) return;
		body.innerHTML = this.entityPageHtml(entry);
		wrap.hidden = false;
	},

	closeEntityPage() {
		const wrap = document.getElementById('entity-page');
		if (wrap) wrap.hidden = true;
	},

	/** Delegated open/close wiring — bound once, survives every re-render. */
	bindEntityPage() {
		if (this._entityPageBound) return;
		this._entityPageBound = true;
		const wrap = document.getElementById('entity-page');
		if (wrap) {
			wrap.addEventListener('click', (evt) => {
				if (evt.target.closest('[data-entity-close]')) this.closeEntityPage();
			});
		}
		document.addEventListener('keydown', (evt) => {
			if (evt.key === 'Escape') this.closeEntityPage();
		});
		// The event-detail «entitet →» links live inside the agenda / Nyheter
		// detail markup, so a document-level delegate catches them once, wherever
		// they are re-rendered. `.ev-entity` is a <button>, not an <a>, so the
		// agenda's own row-toggle listener ignores it.
		document.addEventListener('click', (evt) => {
			const link = evt.target.closest && evt.target.closest('.ev-entity[data-entity-key]');
			if (!link) return;
			evt.preventDefault();
			const entry = this.entityForKey ? this.entityForKey(link.dataset.entityKey) : null;
			if (entry) this.openEntityPage(entry);
		});
	},
});

/** Norwegian words for an entities.json `type` (mirrors FollowVocabulary.typeLabel). */
const SS_EP_TYPE_NB = {
	athlete: 'utøver', team: 'lag', tournament: 'turnering',
	league: 'liga', sport: 'sport', category: 'kategori',
};

if (typeof window !== 'undefined') {
	window.ssEntityUpcoming = ssEntityUpcoming;
	window.ssEntityResults = ssEntityResults;
	window.ssEntityNews = ssEntityNews;
	window.ssEntityStandingsTable = ssEntityStandingsTable;
	window.ssSpecialistLink = ssSpecialistLink;
}
if (typeof module !== 'undefined' && module.exports) {
	module.exports = { ssEntityUpcoming, ssEntityResults, ssEntityNews, ssEntityStandingsTable, ssSpecialistLink };
}
